pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -574,35 +580,102 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
+
+		// if the feature was enabled and then later disabled, MaxUnavailable may have a value
+		// other than 1. Ignore the passed in value and Use maxUnavailable as 1 to enforce
+		// expected behavior when feature gate is not enabled.
+		if utilfeature.DefaultFeatureGate.Enabled(features.MaxUnavailableStatefulSet) {
+			maxUnavailable, err = intstrutil.GetValueFromIntOrPercent(intstrutil.ValueOrDefault(set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromInt(1)), int(replicaCount), false)
+			if err != nil {
+				return &status, err
+			}
+
+		}
 	}
-	// we terminate the Pod with the largest ordinal that does not match the update revision.
+
+	fmt.Println("reached here1 with maxUnavailable= ", maxUnavailable)
+	// Collect all targets in the range between the partition and Spec.Replicas. Count any targets in that range that are unhealthy
+	// i.e. terminated or not running and ready as unavailable). Select the (MaxUnavailable - Unavailable) Pods, in order with respect
+	// to their ordinal for termination. Delete those pods and count the successful deletions. Update the status with the correct
+	// number of deletions.
+	unavailablePods := 0
 	for target := len(replicas) - 1; target >= updateMin; target-- {
+		if !isHealthy(replicas[target]) {
+			unavailablePods++
+		}
+	}
+
+	fmt.Printf("reached here11 found unavailablePods %d, maxUnavailable=%d\n", unavailablePods, maxUnavailable)
 
-		// delete the Pod if it is not already terminating and does not match the update revision.
+	// Now we need to delete MaxUnavailable- unavailablePods
+	// start deleting one by one starting from the highest ordinal first
+	podsToDelete := maxUnavailable - unavailablePods
+	deletedPods := 0
+	for target := len(replicas) - 1; target >= updateMin && podsToDelete > 0; target-- {
+
+		fmt.Println("reached here2")
+		// delete the Pod if it is healthy and the revision doesnt match the target
 		if getPodRevision(replicas[target]) != updateRevision.Name && !isTerminating(replicas[target]) {
+			fmt.Println("reached here3")
+			// delete the Pod if it is healthy and the revision doesnt match the target
 			klog.V(2).Infof("StatefulSet %s/%s terminating Pod %s for update",
 				set.Namespace,
 				set.Name,
 				replicas[target].Name)
-			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			if err := ssc.podControl.DeleteStatefulPod(set, replicas[target]); err != nil {
+				if !errors.IsNotFound(err) {
+					return nil, err
+				}
+			}
+			deletedPods++
+			fmt.Println("reached here4")
 			status.CurrentReplicas--
-			return &status, err
 		}
 
-		// wait for unhealthy Pods on update
-		if !isHealthy(replicas[target]) {
-			klog.V(4).Infof(
-				"StatefulSet %s/%s is waiting for Pod %s to update",
-				set.Namespace,
-				set.Name,
-				replicas[target].Name)
+		// If at anytime, total number of unavailable Pods exceeds maxUnavailable,
+		// we stop deleting more Pods for Update
+		if deletedPods == podsToDelete {
```

Comment:
I think we also need to check based on `updateMin`

Reply:
that is already part of the for loop, see line 645 which starts the for loop

Reply:
What I meant was this can be simplified if have stuck to updateMin and updated that var itself.

Reply:
updateMin is not a count, its the ordinal of the pod from which deletion needs to begin based on partition. @ravisantoshgudimetla